File: app/src/main/java/com/galib/natorepbs2/db/NPBS2DB.java
package com.galib.natorepbs2.db;

import android.content.Context;

import androidx.room.Database;
import androidx.room.Room;
import androidx.room.RoomDatabase;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

@Database(entities = {Information.class, Achievement.class, ComplainCentre.class, Employee.class, OfficeInformation.class, TenderInformation.class}, version = 12, exportSchema = false)
public abstract class NPBS2DB extends RoomDatabase {

    public abstract InformationDao informationDao();
    public abstract AchievementDao achievementDao();
    public abstract ComplainCentreDao complainCentreDao();
    public abstract EmployeeDao employeeDao();
    public abstract OfficeInformationDao officeInformationDao();
    public abstract TenderInformationDao tenderInformationDao();

    private static volatile NPBS2DB INSTANCE;
    private static final int NUMBER_OF_THREADS = 4;
    static final ExecutorService databaseWriteExecutor =
            Executors.newFixedThreadPool(NUMBER_OF_THREADS);

    static NPBS2DB getDatabase(final Context context) {
        if (INSTANCE == null) {
            synchronized (NPBS2DB.class) {
                if (INSTANCE == null) {
                    INSTANCE = Room.databaseBuilder(context.getApplicationContext(),
                                    NPBS2DB.class, "npbs2_database")
                            .fallbackToDestructiveMigration()
                            .build();
                }
            }
        }
        return INSTANCE;
    }
}

File: app/src/main/java/com/galib/natorepbs2/db/NPBS2DB.kt
package com.galib.natorepbs2.db

import android.content.Context
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase
import com.galib.natorepbs2.models.AccountByCC
import com.galib.natorepbs2.models.Achievement
import com.galib.natorepbs2.models.ComplainCentre
import com.galib.natorepbs2.models.Employee
import com.galib.natorepbs2.models.Information
import com.galib.natorepbs2.models.MyMenuItem
import com.galib.natorepbs2.models.NoticeInformation
import com.galib.natorepbs2.models.OfficeInformation

@Database(
    entities = [Information::class, Achievement::class, ComplainCentre::class, AccountByCC::class, Employee::class, OfficeInformation::class, NoticeInformation::class, MyMenuItem::class],
    version = 4,
    exportSchema = false
)
abstract class NPBS2DB : RoomDatabase() {
    abstract fun informationDao(): InformationDao
    abstract fun achievementDao(): AchievementDao
    abstract fun complainCentreDao(): ComplainCentreDao
    abstract fun accountByCCDao(): AccountByCCDao
    abstract fun employeeDao(): EmployeeDao
    abstract fun officeInformationDao(): OfficeInformationDao
    abstract fun noticeInformationDao(): NoticeInformationDao
    abstract fun myMenuItemDao(): MyMenuItemDao

    companion object {
        @Volatile
        private var INSTANCE: NPBS2DB? = null
        fun getDatabase(context: Context): NPBS2DB {
            return INSTANCE ?: synchronized(this) {
                val instance = Room.databaseBuilder(
                    context.applicationContext,
                    NPBS2DB::class.java,
                    "npbs2_database"
                )
                    .createFromAsset("database/npbs2_database") // db file database version must be same as current db version
//                    .setQueryCallback({ sqlQuery, bindArgs ->
//                        Log.d("NPBS2DB", "SQL Query: $sqlQuery SQL Args: $bindArgs")
//                    }, Executors.newSingleThreadExecutor())
                    .fallbackToDestructiveMigration()
                    .build()
                INSTANCE = instance
                // return instance
                instance
            }
        }
    }
}


